thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/anomaly/events/HolidayEventsLoader.java

```diff
@@ -0,0 +1,337 @@
+    }
+  }
+
+  private static final DAORegistry DAO_REGISTRY = DAORegistry.getInstance();
+
+  /**
+   * Start.
+   */
+  public void start() {
+    scheduledExecutorService.scheduleAtFixedRate(this, 0, runFrequency.getSize(), runFrequency.getUnit());
+  }
+
+  /**
+   * Shutdown.
+   */
+  public void shutdown() {
+    scheduledExecutorService.shutdown();
+  }
+
+  /**
+   * Fetch holidays and save to ThirdEye database.
+   */
+  public void run() {
+    EventManager eventDAO = DAO_REGISTRY.getEventDAO();
+
+    long start = System.currentTimeMillis();
+    long end = start + 2592000000L; // to get holidays within a month
+
+    List<Event> newHolidays = getAllHolidays(start, end);
+
+    // A map from new holiday to a set of country codes that has the holiday
+    Map<HolidayEvent, Set<String>> newHolidayEventToCountryCodes = new HashMap<>();
+
+    // Group by holidays and put the country codes in the corresponding set
+    for (Event holiday : newHolidays) {
+      HolidayEvent holidayEvent =
+          new HolidayEvent(holiday.getSummary(), EventType.HOLIDAY.toString(), holiday.getStart().getDate().getValue(),
+              holiday.getEnd().getDate().getValue());
+
+      if (!newHolidayEventToCountryCodes.containsKey(holidayEvent)) {
+        newHolidayEventToCountryCodes.put(holidayEvent, new HashSet<String>());
+      }
+      newHolidayEventToCountryCodes.get(holidayEvent).add(getCountryCode(holiday));
+    }
+
+    // Get the existing holidays in the time range
+    List<EventDTO> existingEvents = eventDAO.findEventsBetweenTimeRange(EventType.HOLIDAY.toString(), start, end);
+    Map<HolidayEvent, EventDTO> existingHolidayEventToEventDTO = new HashMap<>();
+    Map<HolidayEvent, Set<String>> existingHolidayEventToCountryCodes = new HashMap<>();
+
+    // Map existing event DTO to the tuple of HolidayEvent and country code set
+    for (EventDTO existingEventDTO : existingEvents) {
+      HolidayEvent holidayEvent = new HolidayEvent(existingEventDTO);
+      if (!existingHolidayEventToCountryCodes.containsKey(holidayEvent)) {
+        existingHolidayEventToCountryCodes.put(holidayEvent, new HashSet<String>());
+      }
+      if (existingEventDTO.getTargetDimensionMap() != null
+          && existingEventDTO.getTargetDimensionMap().get("countryCode") != null) {
+        existingHolidayEventToCountryCodes.get(holidayEvent)
+            .addAll(existingEventDTO.getTargetDimensionMap().get("countryCode"));
+      }
+      existingHolidayEventToEventDTO.put(holidayEvent, existingEventDTO);
+    }
+
+    for (Map.Entry<HolidayEvent, Set<String>> entry : newHolidayEventToCountryCodes.entrySet()) {
+      HolidayEvent newHolidayEvent = entry.getKey();
+      Set<String> newCountryCodes = entry.getValue();
+
+      if (existingHolidayEventToCountryCodes.containsKey(newHolidayEvent)) {
+        if (existingHolidayEventToCountryCodes.get(newHolidayEvent).addAll(newCountryCodes)) {
```

Comment:
Take care of the duplicated country codes in existingHolidayEventToCountryCodes.get(newHolidayEvent) and newCountryCodes. You could insert all country codes to a set and output the unique code to the list.

Reply:
existingHolidayEventToCountryCodes maps the holiday in to a country codes hashset.  So this shouldn't be a problem